File: kotlin/src/main/kotlin/aoc/common/Answer.kt
package aoc.common

private fun formatPrefix(year: Int, day: Int, part: Int? = null): String {
    val dayStr = day.toString().padStart(2, '0')
    val partStr = part?.let { " [Part $it]" }.orEmpty()
    return "[AOC $year] [Day $dayStr]$partStr"
}

fun <T> answer(year: Int, day: Int, part: Int? = null, result: T) {
    println("${formatPrefix(year, day, part)} $result")
}

fun <T> answer(year: Int, day: Int, result: T) {
    answer(year, day, null, result)
}

fun <T> verify(year: Int, day: Int, part: Int? = null, expected: T, actual: T) {
    val status = if (expected == actual) "PASS" else "FAIL (expected: $expected, actual: $actual)"
    println("${formatPrefix(year, day, part)} $status")
}

fun <T> verify(year: Int, day: Int, expected: T, actual: T) {
    verify(year, day, null, expected, actual)
}

File: typescript/src/common/answer.ts
interface AocContext {
  year: number;
  day: number;
  part?: number;
}

function formatPrefix({ year, day, part }: AocContext): string {
  const dayStr = String(day).padStart(2, "0");
  const partStr = part ? ` [Part ${part}]` : "";
  return `[AOC ${year}] [Day ${dayStr}]${partStr}`;
}

export function answer<T>(ctx: AocContext, result: T): void {
  console.log(`${formatPrefix(ctx)} ${result}`);
}

export function verify<T>(ctx: AocContext, expected: T, actual: T): void {
  const status =
    expected === actual
      ? "PASS"
      : `FAIL (expected: ${expected}, actual: ${actual})`;
  console.log(`${formatPrefix(ctx)} ${status}`);
}